src/core/util/autofocus.js

```diff
@@ -0,0 +1,105 @@
+  .directive('mdAutoFocus', MdAutofocusDirective)
+  .directive('mdSidenavFocus', MdAutofocusDirective);
+
+/**
+ * @ngdoc directive
+ * @name mdAutofocus
+ * @module material.core.util
+ *
+
+ *
+ * @description
+ * `$mdUtil.findFocusTarget()` provides an optional way to identify the focused element when a dialog, bottomsheet, sideNav
+ * or other element opens. This is optional attribute finds a nested element with the mdAutoFocus attribute and optional
+ * expression. An expression may be specified as the directive value; to enable conditional activation of the autoFocus.
+ *
+ * @usage
+ * ### Dialog
+ * <hljs lang="html">
+ * <md-dialog>
+ *   <form>
+ *     <md-input-container>
+ *       <label for="testInput">Label</label>
+ *       <input id="testInput" type="text" md-autofocus>
+ *     </md-input-container>
+ *   </form>
+ * </md-dialog>
+ * </hljs>
+ *
+ * ### Bottomsheet
+ * <hljs lang="html">
+ * <md-bottom-sheet class="md-list md-has-header">
+ *  <md-subheader>Comment Actions</md-subheader>
+ *  <md-list>
+ *    <md-list-item ng-repeat="item in items">
+ *
+ *      <md-button md-autofocus="$index == 2">
+ *        <md-icon md-svg-src="{{item.icon}}"></md-icon>
+ *        <span class="md-inline-list-icon-label">{{ item.name }}</span>
+ *      </md-button>
+ *
+ *    </md-list-item>
+ *  </md-list>
+ * </md-bottom-sheet>
+ * </hljs>
+ *
+ * ### Autocomplete
+ * <hljs lang="html">
+ *   <md-autocomplete
+ *       md-autofocus
+ *       md-selected-item="selectedItem"
+ *       md-search-text="searchText"
+ *       md-items="item in getMatches(searchText)"
+ *       md-item-text="item.display">
+ *     <span md-highlight-text="searchText">{{item.display}}</span>
+ *   </md-autocomplete>
+ * </hljs>
+ *
+ * ### Sidenav
+ * <hljs lang="html">
+ * <div layout="row" ng-controller="MyController">
+ *   <md-sidenav md-component-id="left" class="md-sidenav-left">
+ *     Left Nav!
+ *   </md-sidenav>
+ *
+ *   <md-content>
+ *     Center Content
+ *     <md-button ng-click="openLeftMenu()">
+ *       Open Left Menu
+ *     </md-button>
+ *   </md-content>
+ *
+ *   <md-sidenav md-component-id="right"
+ *     md-is-locked-open="$mdMedia('min-width: 333px')"
+ *     class="md-sidenav-right">
+ *     <form>
+ *       <md-input-container>
+ *         <label for="testInput">Test input</label>
+ *         <input id="testInput" type="text"
+ *                ng-model="data" md-autofocus>
+ *       </md-input-container>
+ *     </form>
+ *   </md-sidenav>
+ * </div>
+ * </hljs>
+ **/
+function MdAutofocusDirective() {
+  return {
+    restrict: 'A',
+
+    link: postLink
+  }
+}
+
+function postLink(scope, element, attrs) {
+  var attr = attrs.mdAutoFocus || attrs.mdAutofocus || attrs.mdSidenavFocus;
+
+  // Setup a watcher on the proper attribute to update a class we can check for in $mdUtil
+  scope.$watch(attr, function(canAutofocus) {
+    element.toggleClass('_md-can-autofocus', canAutofocus);
```

Comment:
`attrs.$observe()` is used for interpolated values; we're not using interpolation here, so I think `$watch` is better: (see https://docs.angularjs.org/api/ng/type/$compile.directive.Attributes and http://stackoverflow.com/questions/14876112/angularjs-difference-between-the-observe-and-watch-methods for more info; I had to re-read them to check)

Basically, the value we get from using `$observe()` is the string `"shouldAutoFocus==1"` which we would then have to evaluate and setup a watcher for. I guess we don't cover the case of someone changing the actual expression (like making it `shouldAutoFocus==2`), but I doubt that people will do that.

I'm adding some additional tests to ensure it works for all types of expressions.

Regarding the `md-autofocus`, I'm not opposed to renaming it (I don't like "can" either), however, I can't think of a reason to make it public (instead of private). This class would be applied to any element that "can" automatically receive focus, not one that currently has focus (I imagine it will get `md-focused` in that case). So, I don't know why people would style it unless they want to make all autofocus-capable inputs stand out a bit.

If we don't make it public, then we can indeed set it on `element.data` instead of a class; I generally just use classes so developers can see it visually in the inspector pane a bit more easily.